// cypress-terminal-report/src/installLogsCollector.schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstallLogsCollectorSchema = void 0;
const superstruct_1 = require("superstruct");
const InstallLogsCollectorSchema = (0, superstruct_1.object)({
    collectTypes: (0, superstruct_1.optional)((0, superstruct_1.array)((0, superstruct_1.enums)([
        "cons:log",
        "cons:info",
        "cons:warn",
        "cons:error",
        "cons:debug",
        "cy:log",
        "cy:xhr",
        "cy:fetch",
        "cy:request",
        "cy:intercept",
        "cy:command"
    ]))),
    filterLog: (0, superstruct_1.optional)((0, superstruct_1.func)()),
    processLog: (0, superstruct_1.optional)((0, superstruct_1.func)()),
    collectTestLogs: (0, superstruct_1.optional)((0, superstruct_1.func)()),
    xhr: (0, superstruct_1.optional)((0, superstruct_1.object)({
        printHeaderData: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
        printRequestData: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
        printBody: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
    })),
    enableExtendedCollector: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
    enableContinuousLogging: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
    commandTimings: (0, superstruct_1.optional)((0, superstruct_1.enums)(["timestamp", "seconds"])),
    debug: (0, superstruct_1.optional)((0, superstruct_1.boolean)()),
});
exports.InstallLogsCollectorSchema = InstallLogsCollectorSchema;
